x-pack/plugin/logstash/src/main/java/org/elasticsearch/xpack/logstash/action/TransportGetPipelineAction.java

```diff
@@ -118,37 +137,64 @@ private void handleFailure(Exception e, ActionListener<GetPipelineResponse> list
         }
     }
 
-    private void handleSearchResponse(
+    private void handleFilteringSearchResponse(
         SearchResponse searchResponse,
         Map<String, BytesReference> pipelineSources,
+        Set<String> explicitPipelineIds,
+        Set<Pattern> wildcardPipelinePatterns,
+        int numberOfHitsSeenPreviously,
         Consumer<SearchResponse> clearScroll,
         ActionListener<GetPipelineResponse> listener
     ) {
-        for (SearchHit hit : searchResponse.getHits().getHits()) {
-            pipelineSources.put(hit.getId(), hit.getSourceRef());
-        }
-
-        if (pipelineSources.size() > searchResponse.getHits().getTotalHits().value) {
+        int numberOfHitsSeenSoFar = numberOfHitsSeenPreviously + searchResponse.getHits().getHits().length;
+        if (numberOfHitsSeenSoFar > searchResponse.getHits().getTotalHits().value) {
             clearScroll.accept(searchResponse);
             listener.onFailure(
                 new IllegalStateException(
                     "scrolling returned more hits ["
-                        + pipelineSources.size()
+                        + numberOfHitsSeenSoFar
                         + "] than expected ["
                         + searchResponse.getHits().getTotalHits().value
                         + "] so bailing out to prevent unbounded "
                         + "memory consumption."
                 )
             );
-        } else if (pipelineSources.size() == searchResponse.getHits().getTotalHits().value) {
+        }
+
+        for (SearchHit hit : searchResponse.getHits().getHits()) {
+            if (explicitPipelineIds.isEmpty() && wildcardPipelinePatterns.isEmpty()) {
+                pipelineSources.put(hit.getId(), hit.getSourceRef());
+                continue;
+            }
+
+            // take if ID is in request IDs set
+            if (explicitPipelineIds.contains(hit.getId())) {
+                pipelineSources.put(hit.getId(), hit.getSourceRef());
+                continue;
+            }
+            // take if id matches request wildcard pattern
+            if (wildcardPipelinePatterns.stream().anyMatch(pattern -> pattern.matcher(hit.getId()).matches())) {
+                pipelineSources.put(hit.getId(), hit.getSourceRef());
+            }
```

Comment:
There's a lot of repetition here - Each of these conditionals has the same body - can we combine the conditions together into one boolean?